manager/assets/modext/widgets/system/modx.grid.context.js

```diff
@@ -142,68 +151,119 @@ Ext.extend(MODx.grid.Context,MODx.grid.Grid,{
-			    'success'	: {
-			    	fn			: function() {
-						this.afterAction();
-					},
-			    	scope		: this
-			    }
-			}
-		});
-	        
-		this.createWindow.show(e.target);    
+        if (this.createWindow) {
+            this.createWindow.destroy();
+        }
+
+        this.createWindow = MODx.load({
+            xtype		: 'modx-window-context-create',
+            closeAction	:'close',
+            listeners	: {
+                'success'	: {
+                    fn			: function() {
+                        this.afterAction();
+                    },
+                    scope		: this
+                }
+            }
+        });
+
+        this.createWindow.show(e.target);
     }
-    
+
     ,remove: function(btn, e) {
-    	MODx.msg.confirm({
-        	title 		: _('warning'),
-        	text		: _('context_remove_confirm'),
-        	url			: this.config.url,
-        	params		: {
-            	action		: 'context/remove',
-            	key			: this.menu.record.key
+        MODx.msg.confirm({
+            title 		: _('warning'),
+            text		: _('context_remove_confirm'),
+            url			: this.config.url,
+            params		: {
+                action		: 'context/remove',
+                key			: this.menu.record.key
             },
             listeners	: {
-            	'success'	: {
-            		fn			: function() {
-	            		this.afterAction();
-            		},
-            		scope		: this
-            	}
+                'success'	: {
+                    fn			: function() {
+                        this.afterAction();
+                    },
+                    scope		: this
+                }
             }
-    	});
+        });
     }
-    
+
     ,afterAction: function() {
-	    var cmp = Ext.getCmp('modx-resource-tree');
-		
-		if (cmp) {
-    		cmp.refresh();
-		}
-		
-		this.getSelectionModel().clearSelections(true);
-		
-		this.refresh();
+        var cmp = Ext.getCmp('modx-resource-tree');
+
+        if (cmp) {
+            cmp.refresh();
+        }
+
+        this.getSelectionModel().clearSelections(true);
+
+        this.refresh();
     }
 
+    ,buttonColumnRenderer: function (a,id,row,ri) {
+        var p = row.data.perm;
+        var b = [];
+        if (p.indexOf('pedit') != -1) {
+            b.push({
+                className: 'update',
+                icon: 'pencil-square-o',
+                text: _('context_update')
+            });
+        }
+        if (p.indexOf('pnew') != -1) {
+            b.push({
+                className: 'duplicate',
+                icon: 'files-o',
+                text: _('context_duplicate')
+            });
+        }
+        if (p.indexOf('premove') != -1) {
+            b.push({
+                className: 'remove',
+                icon: 'trash-o',
+                text: _('context_remove')
+            });
+        }
+        return this.getButtonColumnTpl().apply({
+            action_buttons: b
+        });
+    },
+    onClick: function (e) {
```

Comment:
I've not tested it yet, but I wonder, could we get rid of the onClick here if each of the action buttons is assigned a `handler` instead? 

Maybe not because they're not extjs-type buttons.. but worth a check to make the per-grid implementation easier perhaps?

Reply:
In my opinion the `handler` does not work in this case, but I'm not sure.
By the way, for packages, the code works similarly - https://github.com/modxcms/revolution/blob/3.x/manager/assets/modext/workspace/package.grid.js#L234

Reply:
It should be more effective the current way, since each button would need a click handler (or listener) then. At the moment the whole button column gets one click handler assigned.